[addons] add GetTime... function and possibility to get ID3D11DeviceContext all the time

AlwinEsch opened the pull request:
## Description

Commit 1:
-----------------------------------------------
**[addons]** add possibility to get ID3D11DeviceContext all the time

Have seen by help of @afedchin that the context can be change on the fly.
Before was to Screensaver and Visualization this only given during start.

This adds a new addon interface function (`void* kodi::gui::GetHWContext()`)
where the addon can ask about them if needed.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

Commit 2:
-----------------------------------------------
**[addons]** this adds helper functions to get current time

The code happens on [p8platform](https://github.com/Pulse-Eight/platform/blob/master/src/util/timeutils.h#L78), but since this defines "long"
as value, it is useless because it is too inaccurate in frames.

Since a lot of addons (over 20) need this I think it belongs to the dev kit.

The code is completely independent of Kodi and gives no access to it.

## Motivation and Context
Improve and make DirectX more safe
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
On screensaver addons
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

afedchin reviewed (changes requested):
to avoid a lot of ifdefs let's introduce a method in windowing `GetHWContext` and use it via service broker. on windows windowing has pointer to `DeviceResources`

afedchin commented on xbmc/windowing/win10/WinSystemWin10.h:
```diff
@@ -101,6 +101,7 @@ class CWinSystemWin10 : public CWinSystemBase
 
 protected:
   bool CreateNewWindow(const std::string& name, bool fullScreen, RESOLUTION_INFO& res) override = 0;
+  void* GetHWContext() override = 0;
```
I see no reason to add this.

AlwinEsch replied:
Seen it on the other and thought to do the same but you are right, by `CreateNewWindow` the CWinSystemBase is also defined with `=0` and then makes it sense.

afedchin commented on xbmc/windowing/windows/WinSystemWin32.h:
```diff
@@ -128,6 +128,7 @@ class CWinSystemWin32 : public CWinSystemBase
 
 protected:
   bool CreateNewWindow(const std::string& name, bool fullScreen, RESOLUTION_INFO& res) override = 0;
+  void* GetHWContext() override = 0;
```
same as above

AlwinEsch commented:
How is the current merge window, can be this in or wait?

jenkins build this please

afedchin commented:
wait until Leia will being branched

Rechi commented:
wrong revision was build
jenkins build this please

garbear commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h:
```diff
@@ -141,5 +141,35 @@ namespace gui
+  /// ~~~~~~~~~~~~~{.cpp}
+  /// #include <d3d11_1.h>
+  /// ..
+  /// ID3D11DeviceContext1* context = reinterpret_cast<ID3D11DeviceContext1>(kodi::gui::GetD3DContext());
```
This line doesn't look right

AlwinEsch replied:
Ups, becomes changed to `GetHWContext` (Was the name before)

afedchin replied:
```suggestion
  /// ID3D11DeviceContext1* context = reinterpret_cast<ID3D11DeviceContext1*>(kodi::gui::GetHWContext());
```
wrong reinterpret_cast's type param :)

garbear replied:
Also static_cast might suffice?

AlwinEsch replied:
Is updated, yes with static_cast also worked and is changed.

AlwinEsch committed:
[addons] add possibility to get ID3D11DeviceContext all the time

Have seen by help of afedchin that the context can be change on the fly.
Before was to Screensaver and Visualization this only given during start.

This adds a new addon interface function (`void* kodi::gui::GetHWContext()`)
where the addon can ask about them if needed.

Further is the use of ifdef's removed and becomes handled by WinSystem.h.

AlwinEsch commented:
jenkins build this please

OK to bring in?

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h:
```diff
@@ -0,0 +1,90 @@
+
+namespace kodi
+{
+  namespace time
```
namespaces are not indented, https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h:
```diff
@@ -0,0 +1,90 @@
+    inline double GetTimeMs()
+    {
+    #if defined(TARGET_DARWIN)
+      return (double) (CVGetCurrentHostTime() / (double)(CVGetHostClockFrequency() * 0.001));
```
C++ style casts

AlwinEsch replied:
I look to change them here and on other places, was taken like them from p8platform.

Can someone quikly test that on Mac a:
```cpp
return static_cast<double>(CVGetCurrentHostTime() / static_cast<double>(CVGetHostClockFrequency() * 0.001));
```
works correct.

AlwinEsch replied:
The code is updated and tested on Windows and Linux, only not Mac, think it works but not confirmed!

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h:
```diff
@@ -0,0 +1,90 @@
+      if (QueryPerformanceFrequency(&tickPerSecond))
+      {
+        QueryPerformanceCounter(&tick);
+        return (double) (tick.QuadPart / (tickPerSecond.QuadPart / 1000.));
```
C++ style cast

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h:
```diff
@@ -0,0 +1,90 @@
+    #else
+      timespec time;
+      clock_gettime(CLOCK_MONOTONIC, &time);
+      return (double)time.tv_sec * 1000.0 + time.tv_nsec / 1000000.0;
```
C++ style cast

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h:
```diff
@@ -0,0 +1,90 @@
+    template <class T>
+    inline T GetTimeSec()
+    {
+      return (T)GetTimeMs() / (T)1000.0;
```
C++ style casts

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h:
```diff
@@ -44,7 +44,7 @@
 #define ADDON_GLOBAL_VERSION_GENERAL_DEPENDS          "General.h"
 
-#define ADDON_GLOBAL_VERSION_GUI                      "5.12.0"
+#define ADDON_GLOBAL_VERSION_GUI                      "5.12.1"
```
Is this a compatible API change, means, addons must not be recompiled?

AlwinEsch replied:
Yes, it add only the header for time and for the DirectX added function is on a by `malloc` created structure entry on end.  This is then a bit bigger but becomes never edited on addon side.

https://github.com/xbmc/xbmc/pull/15609/files/a51776e52d7301a7de8e39a28b1b7efdffd7b6c1#diff-c244971ddb8feeb77e6e8a8c7af39a99R29

ksooo replied:
Key for compatibility is to add the new function table entry at the end of the table, as you did. All fine, then.

AlwinEsch committed:
[addons] this adds helper functions to get current time

The code happens on "p8platform", but since this defines "long"
as value, it is useless because it is too inaccurate in frames.

Since a lot of addons (over 20) need this I think it belongs to the dev kit.

The code is completely independent of Kodi and gives no access to it.

AlwinEsch commented:
jenkins build this please
